core/templates/pages/admin-page/activities-tab/admin-dev-mode-activities-tab.component.spec.ts

```diff
@@ -0,0 +1,453 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the admin dev activities tab page.
+ */
+
+import { HttpClientTestingModule } from '@angular/common/http/testing';
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { FormsModule } from '@angular/forms';
+import { MaterialModule } from 'components/material.module';
+import { AdminBackendApiService, AdminPageData } from 'domain/admin/admin-backend-api.service';
+import { WindowRef } from 'services/contextual/window-ref.service';
+import { AdminDataService } from '../services/admin-data.service';
+import { AdminTaskManagerService } from '../services/admin-task-manager.service';
+import { AdminDevModeActivitiesTabComponent } from './admin-dev-mode-activities-tab.component';
+
+let loadNewStructuresData: boolean = true;
+let generateNewSkillData: boolean = true;
+
+class MockAdminBackendApiService {
+  private reloadExplorationAsync(explorationId: string) {
+    return {
+      then: (
+          succesCallback: () => void,
+          errorCallback: (err: object) => void
+      ) => {
+        if (explorationId === 'expId') {
+          succesCallback();
+        }
+        if (errorCallback) {
+          errorCallback({
+            data: {
+              error: 'Exploration not found.'
+            }
+          });
+        }
+      }
+    };
+  }
+
+  private generateDummyExplorationsAsync(
+      numDummyExpsToGenerate: number, numDummyExpsToPublish: number) {
+    return {
+      then: (
+          succesCallback: () => void,
+          errorCallback: (err: object) => void
+      ) => {
+        if (numDummyExpsToGenerate === 2 && numDummyExpsToPublish === 1) {
```

Comment:
Instead of using this if else in the test block should we use spyOn...returnValue in the test book when it's needed? [not sure!]

Reply:
I tried doing that, but was not able to figure out how to handle the 2 callbacks of ```.then```.

Reply:
Where do we have 2 callbacks? [We cannot help until you explain your problem clearly with all the details.]

Reply:
Hey @DubeySandeep I was talking about the successCallback and failureCallback. Although, I tried spying on the service instead of creating a mock service.

For example, the test for generateDummyExplorations()'s failure would look like this:

```
it('should show error message when dummy explorations' +
      'are not generated', async(() => {
      component.numDummyExpsToPublish = 2;
      component.numDummyExpsToGenerate = 2;

      spyOn(adminBackendApiService, 'generateDummyExplorationsAsync')
        .and.returnValue(Promise.reject(
          {
            data: {
              error: "Dummy explorations not generated."
            }
          }
        ));
      spyOn(component.setStatusMessage, 'emit');

      component.generateDummyExplorations();

      fixture.whenStable().then(() => {
        expect(component.setStatusMessage.emit).toHaveBeenCalledWith(
          'Server error: Dummy explorations not generated.');
        expect(adminBackendApiService.generateDummyExplorationsAsync)
          .toHaveBeenCalledWith(2,2);
      });
}));
```

Does this look good?

Reply:
@DubeySandeep nudge

Reply:
The[ above test](https://github.com/oppia/oppia/pull/12189#discussion_r595841297) looks good to me, is there any issue with that?

Reply:
Nope, made the required changes.